File: hackerrank/problem solving/algorithms/2 - implementation/47 - minimum distances/minimum_distances.py
# https://www.hackerrank.com/challenges/minimum-distances/problem?isFullScreen=true

def main():
    input()
    arr = read_numbers()
    print(min_distance(arr))


def read_numbers():
    return list(map(int, input().split()))


def min_distance(arr):
    last_seen = {}
    min_dist = None

    for i, x in enumerate(arr):
        if x in last_seen:
            dist = i - last_seen[x]
            if min_dist is None or dist < min_dist:
                min_dist = dist
                if min_dist == 1:
                    return 1

        last_seen[x] = i

    return -1 if min_dist is None else min_dist


if __name__ == '__main__':
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/47 - minimum distances/minimum_distances.ts
// https://www.hackerrank.com/challenges/minimum-distances/problem?isFullScreen=true

function main() {
  readLine();
  const arr: number[] = readNumbers();
  console.log(minDistance(arr));
}

function readNumbers(): number[] {
  return readLine()
    .split(' ')
    .map((x) => +x);
}

// n: length of arr
// T: O(n)
// S: O(n) extra space
function minDistance(arr: number[]): number {
  const lastSeen: { [key: number]: number } = {};
  let minDist = -1;

  for (let i = 0; i < arr.length; i++) {
    const x = arr[i];

    if (lastSeen.hasOwnProperty(x)) {
      const dist = i - lastSeen[x];

      if (minDist === -1 || dist < minDist) {
        minDist = dist;
        if (minDist === 1) return 1;
      }
    }

    lastSeen[x] = i;
  }

  return minDist;
}

//////////////////////////////////////////////////

import readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
let inputLines: string[] = [];
rl.on('line', (line: string) => inputLines.push(line));
rl.on('close', main);

function readLine(): string {
  return inputLines.shift()!;
}
